include/bitops.h

```diff
@@ -0,0 +1,224 @@
+//
+// Checking:
+//   are_bits_set(reg, bit_6); // true
+//   are_bits_set(reg, bit_7); // true
+//   are_bits_set(reg, bit_8); // ASSERT FAIL (out of range)
+//   are_bits_set(reg, bit_6 | bit_7); // true (both need to be set)
+//   are_bits_set(reg, bit_5 | bit_6); // false (both aren't set)
+//   are_any_bits_set(reg, bit_5 | bit_6); // true (bit 6 is)
+//   are_bits_cleared(reg, bit_0 | bit_1 | bit_2 | bit_3); // true
+//
+
+enum BITS {
+	bit_0 = 1 << 0,
+	bit_1 = 1 << 1,
+	bit_2 = 1 << 2,
+	bit_3 = 1 << 3,
+	bit_4 = 1 << 4,
+	bit_5 = 1 << 5,
+	bit_6 = 1 << 6,
+	bit_7 = 1 << 7,
+	bit_8 = 1 << 8,
+	bit_9 = 1 << 9,
+	bit_10 = 1 << 10,
+	bit_11 = 1 << 11,
+	bit_12 = 1 << 12,
+	bit_13 = 1 << 13,
+	bit_14 = 1 << 14,
+	bit_15 = 1 << 15,
+	bit_16 = 1 << 16,
+	bit_17 = 1 << 17,
+	bit_18 = 1 << 18,
+	bit_19 = 1 << 19,
+	bit_20 = 1 << 20,
+	bit_21 = 1 << 21,
+	bit_22 = 1 << 22,
+	bit_23 = 1 << 23,
+	bit_24 = 1 << 24,
+	bit_25 = 1 << 25,
+	bit_26 = 1 << 26,
+	bit_27 = 1 << 27,
+	bit_28 = 1 << 28,
+	bit_29 = 1 << 29,
+	bit_30 = 1 << 30,
+	bit_31 = 1 << 31
+};
+
+// Is the bitmask within the maximum value of the register?
+template <typename T>
+static constexpr void validate_bit_width([[maybe_unused]] const T reg, const int bits)
+{
+	// ensure the register is unsigned
+	static_assert(std::is_unsigned<T>::value, "register must be unsigned");
+
+	// Ensure the bits fall within the type size
+	assert(static_cast<uint64_t>(bits) > 0);
+	assert(static_cast<int64_t>(bits) <=
+	       static_cast<int64_t>(std::numeric_limits<T>::max()));
+}
+
+// Set the indicated bits
+template <typename T>
+constexpr T set_bits_const(const T reg, const int bits)
+{
+	validate_bit_width(reg, bits);
+	return reg | static_cast<T>(bits);
+}
+template <typename T>
+constexpr void set_bits(T &reg, const int bits)
+{
+	reg = set_bits_const(reg, bits);
+}
+
+// Set all of the register's bits high
+//
+// It's often seen in code like this:
+//    unsigned long int flags = ~0; // [1]
+//    unsigned long int flags = 0xffffffff; // [2]
+//
+// But:
+//   [1] isn't portable because it relies on a two's-complement representation.
+//   [2] isn't portable because it assumes 32-bit ints.
+//
+// This function becomes self-documenting: it tells the reader that we intend to
+// use the bits in the register as opposed to being a plain old number or
+// counter.
+//
+template <typename T>
+constexpr T all_bits_set()
+{
+	return std::is_signed<T>::value ? static_cast<T>(-1)
+	                                : std::numeric_limits<T>::max();
+}
+template <typename T>
+constexpr void set_all_bits(T &reg)
+{
+	reg = all_bits_set<T>();
+}
+
+// Clear the indicated bits
+template <typename T>
+constexpr T clear_bits_const(const T reg, const int bits)
+{
+	validate_bit_width(reg, bits);
+	return reg & ~static_cast<T>(bits);
+}
+template <typename T>
+constexpr void clear_bits(T &reg, const int bits)
+{
+	reg = clear_bits_const(reg, bits);
+}
```

Comment:
Again, suggest maybe overloading if possible.

Same for other functions below.

Reply:
It unfortunately can't be overloaded without causing compiler errors (ambiguous functions).

Reply:
I think our only options involve differentiating based on name. 
Maybe: `const_clear_bits`, `const_set_bits`, `const_toggle_bits`?
or `ro_`, readonly_`, `not_inplace_`, `passthrough_`, `returns_`.. other adjectives?

Reply:
New folding version inbound.
 - Modify-in-place (no return): `set`, `clear`, `flip`
 - Read-only masking operations (with return): `mask_on`, `mask_off`, `mask_flip`